src/Microsoft.VisualStudio.Editors/PropPages/CodeAnalysisPropPage.vb

```diff
@@ -0,0 +1,246 @@
+
+            RefreshInstalledFxCopAnalyzersVersionAndButtons()
+        End Sub
+
+        Protected Overrides ReadOnly Property ControlData() As PropertyControlData()
+            Get
+                If m_ControlData Is Nothing Then
+                    m_ControlData = New PropertyControlData() {
+                        New PropertyControlData(1, "RunRoslynAnalyzersDuringBuild", RunRoslynAnalyzersDuringBuild, ControlDataFlags.None),
+                        New PropertyControlData(2, "RunRoslynAnalyzersDuringLiveAnalysis", RunRoslynAnalyzersDuringLiveAnalysis, ControlDataFlags.None)
+                    }
+                End If
+                Return m_ControlData
+            End Get
+        End Property
+
+        Protected Overrides Function GetF1HelpKeyword() As String
+            ' TODO: New help keyword
+            Return HelpKeywords.VBProjPropAssemblyInfo
+        End Function
+
+        Private Sub RefreshInstalledFxCopAnalyzersVersionAndButtons()
+            If Not TryGetInstalledFxCopAnalyzersVersion(_installedFxCopAnalyzersVersionString) Then
+                RefreshInstallFxCopAnalyzersButtons(reset:=True)
+                Return
+            End If
+
+            InstalledVersionTextBox.Text = If(_installedFxCopAnalyzersVersionString, My.Resources.Strings.NotInstalledText)
+            RefreshInstallFxCopAnalyzersButtons(reset:=False)
+        End Sub
+
+        Private Sub RefreshInstallFxCopAnalyzersButtons(reset As Boolean)
+            If DTEProject Is Nothing Then
+                reset = True
+            End If
+
+            InstallLatestVersionButton.Enabled = Not reset AndAlso Not HasFxCopAnalyzersInstalled()
+            UninstallAnalyzersButton.Enabled = Not reset AndAlso HasFxCopAnalyzersInstalled()
+            InstallCustomVersionButton.Enabled = Not reset AndAlso _nugetPackage IsNot Nothing
+
+            InstallLatestVersionButton.Visible = Not HasFxCopAnalyzersInstalled()
+            UninstallAnalyzersButton.Visible = Not InstallLatestVersionButton.Visible
+        End Sub
+
+        Private Function HasFxCopAnalyzersInstalled() As Boolean
+            Return Not String.IsNullOrEmpty(_installedFxCopAnalyzersVersionString)
+        End Function
+
+        ' Returns False if attempt to GetInstalledPackages did not succeed
+        Private Function TryGetInstalledFxCopAnalyzersVersion(ByRef versionString As String) As Boolean
+            versionString = Nothing
+
+            If DTEProject Is Nothing Then
+                Return False
+            End If
+
+            Using session = _threadedWaitDialogFactory.StartWaitDialog(My.Resources.Strings.RestoringPackagesMessage)
+                _packageRestorer.RestorePackages(DTEProject)
+
+                session.UserCancellationToken.ThrowIfCancellationRequested()
+
+                ' Add a small delay to ensure GetInstalledPackages invocation below gets updated packages.
+                Thread.Sleep(2000)
```

Comment:
Rather than a thread sleep, can this method be made async, and have the below check implemented with a timeout? Presumably for a project that is already fully restored this is 2 seconds of waiting (on the UI thread?) for no real reason. It doesn't look like this method actually needs to return its value synchronously, it can just enable/disable buttons whenever its finished.

Reply:
Yes, that sounds reasonable. I can make that change, thanks!

Reply:
@davidwengier I gave this a shot but the attempt to set the `Text` field below does not succeed even if we switch back to the UI thread:

```vb
         Task.Run(
                Function()
                    ' Add a small delay to ensure GetInstalledPackages invocation below gets updated packages.
                    Thread.Sleep(2000)

                    Dim installedPackages As IEnumerable(Of IVsPackageMetadata)
                    Try
                        installedPackages = _packageInstallerServices.GetInstalledPackages(DTEProject)
                    Catch
                        RefreshInstallFxCopAnalyzersButtons(reset:=True)
                        Return Tasks.Task.CompletedTask
                    End Try

                    _installedFxCopAnalyzersVersionString = Nothing
                    For Each package As IVsPackageMetadata In installedPackages
                        If String.Equals(package.Id, FxCopAnalyzersPackageId, StringComparison.OrdinalIgnoreCase) OrElse
                           s_childPackageIds.Contains(package.Id) Then
                            _installedFxCopAnalyzersVersionString = package.VersionString
                            Exit For
                        End If
                    Next

                    ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync()
                    ' Below update does not succeed and the line after that to refresh button states is never executed
                    InstalledVersionTextBox.Text = If(_installedFxCopAnalyzersVersionString, My.Resources.Strings.NotInstalledText)
                    RefreshInstallFxCopAnalyzersButtons(reset:=False)
                End Function)
        End Sub
```

I am going to stick with the current approach (note that the delay is inside a threaded wait dialog so the current UI experience is not bad).